### PR TITLE
fix(client): add type to warn misuse of attributes in formControl

### PR DESCRIPTION
**Pull request by Sboonny**
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

While using FormControl as component, I have added `value` attribute to the component, which leads to this error in the console

image

With Ahmad and Oliver help, you can use FormControl, add the attribute and get warned for the need for the other attributes. 

I opened a new PR to reduce the scale of the FormGroup PR.

### CODE REVIEW

**Comment by raisedadead** [comment]
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

**Review by Sembauke — CHANGES_REQUESTED** [review]
This is a somewhat more simplified version that also should work for the union types:
```ts
type Values = {
  value: string;
  onChange?: (event: React.ChangeEvent<FormControlElement>) => void;
  readOnly?: boolean;
} | {
  value?: never;
  onChange?: never;
  readOnly?: never;
}
```

**Comment by Sboonny** [comment]
removed `undefined` from `onChange` and `readonly`, then added conditional typing to have the error back.

image
image
image

**Comment by Sboonny** [comment]
Here is context, I should have added:

The browser doesn't like it when we us use `value` without including `onChange` or `readonly` attributes to the `FormControl`, but TypeScript doesn't warn us about this.

This aims to add types that warn us about the missing `onChange` or `readonly` when using the `value` attribute, in the editor.

**Comment by ahmadabdolsaheb** [comment]
We just need to make sure that both onChange and readOnly cannot be passed when there is a value.

File: tools/ui-components/src/form-control/types.ts [hunk 1]
```diff
@@ -2,20 +2,34 @@ import React from 'react';
...
+    }
+  | {
+      value: string;
+      onChange?: never;
+      readonly: boolean;
+    }
+  | {
+      value: string;
+      onChange: (event: React.ChangeEvent<FormControlElement>) => void;
+      readonly: never;
```
Comment (ahmadabdolsaheb) [inline tools/ui-components/src/form-control/types.ts:19]:
```suggestion
      readonly?: never;
```

**Review by ahmaxed — APPROVED** [review]
Great job here. Just a tiny nitpick.

**Review by Sembauke — APPROVED** [review]

**Pull request merged**